qio/src/org/qcmg/qio/record/RecordReader.java

```diff
@@ -0,0 +1,147 @@
+/**
+ * © Copyright The University of Queensland 2010-2014.
+ * © Copyright QIMR Berghofer Medical Research Institute 2014-2016.
+ *
+ * This code is released under the terms outlined in the included LICENSE file.
+ */
+package org.qcmg.qio.record;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+import org.qcmg.common.util.FileUtils;
+
+
+public abstract class RecordReader<T> implements Closeable, Iterable<T> {
+    public static final int DEFAULT_BUFFER_SIZE = 65536;
+    public static final String DEFAULT_HEADER_PREFIX = null; //no header line
+    public static final Charset DEFAULT_CHARSET = StandardCharsets.UTF_8;
+    
+    protected final File file;
+    //only allow create once
+    protected final BufferedReader bin;
+    protected T next; 
+    
+    protected List<String> headerLines = new ArrayList<>();
+    public RecordReader(final File file) throws IOException { 
+    	this(file, DEFAULT_BUFFER_SIZE);
+    }
+    
+    public RecordReader(final File file, int bufferSize) throws IOException {
+    	this(file, bufferSize, DEFAULT_HEADER_PREFIX, DEFAULT_CHARSET);
+    } 
+    
+    public RecordReader(final File file, CharSequence headerPrefix) throws IOException { 
+    	this(file, DEFAULT_BUFFER_SIZE, headerPrefix, DEFAULT_CHARSET); 
+    }
+      
+    public RecordReader(final File file, int bufferSize, CharSequence headerPrefix, Charset charset) throws IOException {
+
+        this.file = file;
+        boolean isGzip = FileUtils.isInputGZip( file);
+        InputStream inputStream =  (isGzip) ? new GZIPInputStream(new FileInputStream(file)) : new FileInputStream(file);  
+        InputStreamReader streamReader = new InputStreamReader(inputStream, charset);
+        bin = new BufferedReader(streamReader, bufferSize);
+               		
+        String nextLine = readHeaderAndReturnFirstNonHeaderLine(headerPrefix);//bin.readLine();        
+		
+		try {
+			//get first record, set to null for empty file
+			next = nextLine == null? null : getRecord(nextLine);
+		}catch(Exception e) {
```

Comment:
IS there a more specific exception that could be caught here? I am always a little uncomfortable with `catch (Exception e) ` - it should be used as a last resort IMO